File: csharp/Tdcv2/Prng/Seekable.cs
namespace Tdcv2.Prng;

/// <summary>
/// Draws that can be taken for one row without taking them for any other.
/// </summary>
/// <remarks>
/// <para>
/// The in-memory engine walks one generator from the start, so row 900 000's value exists only
/// after the 899 999 before it. That is fine when the whole run is in memory and impossible when
/// it is not.
/// </para>
/// <para>
/// Here each draw is keyed by <c>seed | streamId | index</c>, so a row's values are a function of
/// its own number. Nothing has to be kept, nothing has to be replayed, and a run of any size costs
/// the memory of one row. It is also what lets separate workers each render a slice of the same
/// file and agree at the seams.
/// </para>
/// </remarks>
public static class Seekable
{
    /// <summary>Half of a 32-bit unit in the last place — see <see cref="OpenUnit"/>.</summary>
    private const double HalfUlp = 0.5 / 4294967296.0;

    /// <summary>A generator private to one row of one stream.</summary>
    public static Sfc32 Generator(string seed, string streamId, int index)
    {
        int[] s = Prng.Cyrb128(seed + "|" + streamId + "|" + index.ToString(System.Globalization.CultureInfo.InvariantCulture));
        return new Sfc32(s[0], s[1], s[2], s[3]);
    }

    public static double Next(string seed, string streamId, int index) =>
        Generator(seed, streamId, index).Next();

    /// <summary>An integer in <c>[0, n)</c> for this row.</summary>
    public static int NextInt(string seed, string streamId, int index, int n)
    {
        if (n <= 1)
        {
            return 0;
        }

        return (int)Math.Floor(Next(seed, streamId, index) * n);
    }

    /// <summary>
    /// Nudge a raw draw into the open interval (0,1).
    /// </summary>
    /// <remarks>
    /// sfc32 emits values in <c>[0, 1)</c>, and inverse-CDF sampling takes logarithms — at exactly
    /// zero those are infinite. The shift is about 1e-10 and changes nothing statistically.
    /// </remarks>
    public static double OpenUnit(double u) => Math.Min(1 - HalfUlp, Math.Max(HalfUlp, u + HalfUlp));

    /// <summary><c>count</c> uniforms in (0,1) for one row — what a fixed-draw sampler needs.</summary>
    public static double[] Uniforms(string seed, string streamId, int index, int count)
    {
        Sfc32 gen = Generator(seed, streamId, index);
        var result = new double[count];
        for (int k = 0; k < count; k++)
        {
            result[k] = OpenUnit(gen.Next());
        }

        return result;
    }
}

File: python/src/tdcv2/prng/seekable.py
"""Draws addressed by row rather than taken in order.

An ordinary generator has to be walked: the millionth value costs a million calls. Keying the
generator by ``seed | stream | index`` instead makes a row's values a function of its own number,
which is what lets the streaming engine answer row nine million without having built the first
eight.
"""

from __future__ import annotations

import math
import struct
from functools import lru_cache

from .prng import INITIAL, Sfc32, State, absorb, finish

# Half of a 32-bit unit in the last place — see `open_unit`.
_HALF_ULP = 0.5 / 4294967296.0


@lru_cache(maxsize=1024)
def _prefix(seed: str, stream_id: str) -> State:
    """The hash state after ``seed|stream|``, which every row of a stream shares.

    A config has a handful of streams and a run has millions of rows, so this turns the constant
    part of the seed string — usually two thirds of its characters — into one lookup.
    """
    return absorb(INITIAL, f"{seed}|{stream_id}|")


def generator(seed: str, stream_id: str, index: int) -> Sfc32:
    """A generator private to one row of one stream."""
    a, b, c, d = finish(absorb(_prefix(seed, stream_id), str(index)))
    return Sfc32(a, b, c, d)


def next_value(seed: str, stream_id: str, index: int) -> float:
    return generator(seed, stream_id, index).next()


def next_int(seed: str, stream_id: str, index: int, n: int) -> int:
    """An integer in ``[0, n)`` for this row."""
    if n <= 1:
        return 0
    return math.floor(next_value(seed, stream_id, index) * n)


def open_unit(u: float) -> float:
    """Nudge a draw off both endpoints, into the open interval ``(0, 1)``.

    Several distributions take a logarithm of the draw or of its complement, and an exact 0 or 1
    would make that infinite. Half a unit in the last place is the smallest shift that cannot
    change which value a draw would otherwise have selected.
    """
    return min(1 - _HALF_ULP, max(_HALF_ULP, u + _HALF_ULP))


def uniforms(seed: str, stream_id: str, index: int, count: int) -> list[float]:
    """``count`` uniforms in ``(0, 1)`` for one row — what a fixed-draw sampler needs."""
    gen = generator(seed, stream_id, index)
    return [open_unit(gen.next()) for _ in range(count)]


def _bits_hex(value: float) -> str:
    """A double as the 16 hex digits of its IEEE-754 image."""
    return struct.unpack(">Q", struct.pack(">d", value))[0].to_bytes(8, "big").hex()


def hash_unit(n: float, salt: float) -> float:
    """A deterministic value in [0, 1) from a pair of numbers — `hash(n, salt)`.

    The key is built from the IEEE-754 BIT PATTERNS of the two arguments, not from
    their decimal forms: `salt` is any double, and the shortest decimal spelling of
    a double differs between languages, while those 64 bits are pinned by the
    standard and printing an integer as hex is exact everywhere. The mixing is
    cyrb128 and the stream is sfc32 — the PRNG the rest of TDC already runs on.
    """
    return generator("hash", f"{_bits_hex(n)}|{_bits_hex(salt)}", 0).next()


def noise_unit(t: float, scale: float, salt: float) -> float:
    """Smooth one-dimensional value noise — `noise(t, scale, salt)`.

    A drifting baseline is not three sine waves: modulate those however you like and
    a spectrum still shows three pure tones. Here each lattice point is an
    independent draw and only the interpolation between them is smooth, so the
    spectrum is broad.

    `scale` is the wavelength in rows; `salt` picks the series. The easing is the
    classic smoothstep, u*u*(3-2u), zero at both ends with zero slope, so no corner
    appears where one cell meets the next. The interpolation is a*(1-u) + b*u for
    the same reason `lerp` uses it: the lattice points come out EXACTLY equal to
    `hash` there, so a cell boundary is continuous to the last bit.

    A `scale` of zero divides by zero and the answer is NaN — the same answer
    `sqrt(-1)` gives here.
    """
    # Python RAISES on float division by zero where IEEE-754 — and the other four
    # implementations — return an infinity. Downstream that infinity always ends
    # as NaN, whatever `t` is: floor(inf) is inf and inf - inf is NaN, and 0/0 is
    # NaN to begin with. So return NaN directly rather than let this one language
    # throw where the rest answer. Same rule as `_divide` in expr/evaluate.py.
    if scale == 0:
        return math.nan
    x = t / scale
    cell = math.floor(x)
    u = x - cell
    eased = u * u * (3 - 2 * u)
    a = hash_unit(cell, salt)
    b = hash_unit(cell + 1, salt)
    return a * (1 - eased) + b * eased
